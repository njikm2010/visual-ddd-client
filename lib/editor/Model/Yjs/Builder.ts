import { NoopArray } from '@wakeapp/utils';
import { v4 } from 'uuid';
import { Map as YMap } from 'yjs';

import { NodeYMap, MapTypeRoot } from './NodeYMap';

const ROOT_ID = MapTypeRoot;

/**
 * 创建根节点，需要和 NodeYMap 保持同步
 */
export function createRoot(children: string[] = []) {
  return NodeYMap.fromNodePO({
    id: ROOT_ID,
    parent: undefined,
    children,
    properties: {
      __node_name__: ROOT_ID,
      __node_type__: 'node',
    },
  });
}

/**
 * 数据表述
 */
export namespace EditorYjsBuilder {
  /**
   * ID
   * ROOT 根节点 ID
   * {your input} 会自动生成 uuid
   * [your input] 原始值，将去掉 []
   */
  export type ID = `{${string}}` | `[${string}]`;

  export interface Property {
    [key: string]: any;
  }

  export interface Helper {
    /**
     * 是否已存在指定 id
     * @param name
     */
    hasId(name: string): boolean;

    /**
     * 获取指定标识符的 id, 如果不存在就创建一个
     * 如果 name 为空，就创建一个随机 id
     * @param name
     */
    getOrCreateId(name?: string): string;

    /**
     * 获取指定标识符的 id, 如果不存在就抛出错误
     * @param name
     */
    getId(name: string): string;
  }

  export interface Node {
    id: ID;
    /**
     * 节点名称
     */
    name: string;

    /**
     * 属性
     */
    properties?: Property | ((helper: Helper) => Property);

    /**
     * 子节点
     */
    children?: Node[];
  }

  export type EdgeTerminal =
    | ID
    | {
        cell: ID;
        port?: string;
      };

  export interface Edge {
    /**
     * 节点名称
     */
    name: string;

    /**
     * 来源节点
     */
    source: EdgeTerminal;

    /**
     * 目标节点
     */
    target: EdgeTerminal;

    /**
     * 属性
     */
    properties?: Property | ((helper: Helper) => Property);
  }

  export interface Representation {
    nodes: Node[];
    edges: Edge[];
  }
}

export class EditorYjsBuilderHelper implements EditorYjsBuilder.Helper {
  private map = new Map<string, string>();

  private createId() {
    return v4();
  }

  hasId(name: string): boolean {
    return this.map.has(name);
  }

  getOrCreateId(name?: string | undefined): string {
    if (name == null) {
      return this.createId();
    }

    if (this.map.has(name)) {
      return this.map.get(name)!;
    }

    const id = this.createId();
    this.map.set(name, id);

    return id;
  }

  getId(name: string): string {
    if (this.map.has(name)) {
      return this.map.get(name)!;
    }

    throw new Error('id not found: ' + name);
  }
}

/**
 * 从 JSON 数据表述中快速创建 yjs 数据
 * @param representation
 */
export function buildEditorYjs(representation: EditorYjsBuilder.Representation, datasource: YMap<unknown>) {
  const { nodes, edges } = representation;
  const helper = new EditorYjsBuilderHelper();

  const getId = (id: EditorYjsBuilder.ID, idMustExisted: boolean = false) => {
    if (id.startsWith('{')) {
      const realId = id.slice(1, -1);
      return idMustExisted ? helper.getId(realId) : helper.getOrCreateId(realId);
    } else {
      return id.slice(1, -1);
    }
  };

  const createTerminal = (terminal: EditorYjsBuilder.EdgeTerminal) => {
    if (typeof terminal === 'string') {
      return { cell: getId(terminal, true) };
    }

    const { cell, ...other } = terminal;

    return {
      cell: getId(cell, true),
      ...other,
    };
  };

  const register = (id: string, cell: YMap<unknown>) => {
    datasource.set(id, cell);
  };

  function createNodeMap(node: EditorYjsBuilder.Node, parent: string = ROOT_ID): { id: string; map: YMap<unknown> } {
    const { id, name, properties, children } = node;

    const idValue = getId(id);
    const propertiesValue = typeof properties === 'function' ? properties(helper) : properties;
    const childrenValue = children?.map(c => createNodeMap(c, idValue)) ?? [];

    const nodeYMap = NodeYMap.fromNodePO({
      id: idValue,
      parent: parent,
      children: childrenValue.map(i => i.id),
      properties: {
        __node_name__: name,
        __node_type__: 'node',
        ...propertiesValue,
      },
    });

    const nodeMap = nodeYMap.toYMap();

    register(idValue, nodeMap);

    return {
      id: idValue,
      map: nodeMap,
    };
  }

  function createEdge(edge: EditorYjsBuilder.Edge): { id: string; map: YMap<unknown> } {
    const { source, target, properties, name } = edge;

    const idValue = helper.getOrCreateId();
    const propertiesValue = typeof properties === 'function' ? properties(helper) : properties;
    const nodeYMap = NodeYMap.fromNodePO({
      id: idValue,
      parent: ROOT_ID,
      children: [],
      properties: {
        __node_name__: name,
        __node_type__: 'edge',
        source: createTerminal(source),
        target: createTerminal(target),
        ...propertiesValue,
      },
    });

    const edgeMap = nodeYMap.toYMap();

    register(idValue, edgeMap);

    return {
      id: idValue,
      map: edgeMap,
    };
  }

  const nodeMaps = nodes.map(n => createNodeMap(n));
  const edgeMaps = edges.map(e => createEdge(e));

  const root = createRoot(nodeMaps.map(n => n.id).concat(edgeMaps.map(e => e.id)));

  register(ROOT_ID, root.toYMap());

  return {
    root,
    nodeMaps,
    edgeMaps,
  };
}

export function buildEmptyEditorYjs(datasource: YMap<unknown>) {
  return buildEditorYjs(
    {
      nodes: NoopArray,
      edges: NoopArray,
    },
    datasource
  );
}
