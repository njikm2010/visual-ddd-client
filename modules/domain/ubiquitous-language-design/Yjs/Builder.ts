import { UbiquitousLanguageItem } from '../types';
import { Array as YArray, Map as YMap } from 'yjs';
import { ItemWrapper } from './ItemWrapper';
import { v4 } from 'uuid';

export type UbiquitousLanguageBuilderRepresentationItem = Omit<UbiquitousLanguageItem, 'uuid'> & { uuid?: string };

export function buildUbiquitousLanguageYjs(
  representation: UbiquitousLanguageBuilderRepresentationItem[],
  datasource: YArray<YMap<string>>
) {
  const createItem = (item: UbiquitousLanguageBuilderRepresentationItem) => {
    const map = ItemWrapper.from({ ...item, uuid: item.uuid ?? v4() }).toYMap();
    datasource.push([map]);

    return map;
  };

  return representation.map(createItem);
}
