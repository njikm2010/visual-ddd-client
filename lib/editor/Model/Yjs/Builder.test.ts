import { Doc as YDoc, Map as YMap } from 'yjs';
import * as mockedUuid from 'uuid';
import { createRoot, buildEditorYjs, buildEmptyEditorYjs, EditorYjsBuilderHelper } from './Builder';

jest.mock('uuid', () => {
  let count = 0;
  return {
    __esModule: true,
    reset() {
      count = 0;
    },
    v4() {
      return `uuid-${count++}`;
    },
  };
});

const mockedUuidModule = mockedUuid as unknown as jest.Mocked<{ reset(): void }>;

function createDoc(): [YDoc, YMap<unknown>] {
  const doc = new YDoc();
  const map = doc.getMap('test');

  return [doc, map];
}

test('createRoot', () => {
  const [doc, map] = createDoc();
  const root = createRoot(['1', '2', '3']);

  map.set('root', root.toYMap());

  expect(doc.toJSON()).toMatchSnapshot();
});

describe('buildEditorYjs', () => {
  beforeEach(() => {
    mockedUuidModule.reset();
  });

  test('EditorYjsBuilderHelper', () => {
    const helper = new EditorYjsBuilderHelper();

    expect(helper.getOrCreateId('foo')).toBe('uuid-0');
    expect(helper.getOrCreateId('foo')).toBe('uuid-0');
    expect(helper.getId('foo')).toBe('uuid-0');
    expect(helper.getOrCreateId('bar')).toBe('uuid-1');
    expect(helper.getOrCreateId()).toBe('uuid-2');

    expect(() => {
      helper.getId('baz');
    }).toThrowError('id not found: baz');
  });

  test('initial State', () => {
    const [doc, map] = createDoc();

    buildEditorYjs(
      {
        nodes: [],
        edges: [],
      },
      map
    );

    expect(doc.toJSON()).toMatchSnapshot();

    const anotherMap = doc.getMap('another');
    buildEmptyEditorYjs(anotherMap);
    expect(map.toJSON()).toEqual(anotherMap.toJSON());
  });

  test('with nodes', () => {
    const [doc, map] = createDoc();

    buildEditorYjs(
      {
        nodes: [
          {
            id: '[keep-it]',
            name: 'my-node',
            properties: {
              a: 1,
              b: 2,
            },
          },
          {
            id: '{foo}',
            name: 'foo',
            properties: helper => ({
              b: 1,
              c: 2,
              d: helper.getId('foo'),
            }),
          },
        ],
        edges: [],
      },
      map
    );

    expect(doc.toJSON()).toMatchSnapshot();
  });

  test('with edges', () => {
    const [doc, map] = createDoc();

    buildEditorYjs(
      {
        nodes: [
          {
            id: '{a}',
            name: 'a',
          },
          {
            id: '{b}',
            name: 'b',
          },
        ],
        edges: [
          {
            source: '{a}',
            target: '{b}',
            name: 'my-edge-a',
          },
          {
            source: { cell: '{b}' },
            target: { cell: '[keep-it]', port: 'port-a' },
            name: 'my-edge-b',
          },
        ],
      },
      map
    );

    expect(doc.toJSON()).toMatchSnapshot();
  });

  test('with children', () => {
    const [doc, map] = createDoc();

    buildEditorYjs(
      {
        nodes: [
          {
            id: '{a}',
            name: 'a',
            children: [
              {
                id: '{b}',
                name: 'b',
              },
              {
                id: '{c}',
                name: 'c',
                properties: {
                  foo: 'bar',
                },
                children: [{ id: '{d}', name: 'd' }],
              },
            ],
          },
        ],
        edges: [
          {
            source: '{a}',
            target: '{d}',
            name: 'edge',
          },
        ],
      },
      map
    );

    expect(doc.toJSON()).toMatchSnapshot();
  });
});
