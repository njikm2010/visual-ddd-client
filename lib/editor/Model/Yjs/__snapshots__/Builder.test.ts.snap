// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildEditorYjs initial State 1`] = `
{
  "test": {
    "__ROOT__": {
      "__NODE__": true,
      "children": {},
      "id": "__ROOT__",
      "locked": undefined,
      "parent": undefined,
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "__ROOT__",
        "__node_type__": "node",
      },
    },
  },
}
`;

exports[`buildEditorYjs with children 1`] = `
{
  "test": {
    "__ROOT__": {
      "__NODE__": true,
      "children": {
        "uuid-0": 1,
        "uuid-4": 1,
      },
      "id": "__ROOT__",
      "locked": undefined,
      "parent": undefined,
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "__ROOT__",
        "__node_type__": "node",
      },
    },
    "uuid-0": {
      "__NODE__": true,
      "children": {
        "uuid-1": 1,
        "uuid-2": 1,
      },
      "id": "uuid-0",
      "locked": undefined,
      "parent": "__ROOT__",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "a",
        "__node_type__": "node",
      },
    },
    "uuid-1": {
      "__NODE__": true,
      "children": {},
      "id": "uuid-1",
      "locked": undefined,
      "parent": "uuid-0",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "b",
        "__node_type__": "node",
      },
    },
    "uuid-2": {
      "__NODE__": true,
      "children": {
        "uuid-3": 1,
      },
      "id": "uuid-2",
      "locked": undefined,
      "parent": "uuid-0",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "c",
        "__node_type__": "node",
        "foo": "bar",
      },
    },
    "uuid-3": {
      "__NODE__": true,
      "children": {},
      "id": "uuid-3",
      "locked": undefined,
      "parent": "uuid-2",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "d",
        "__node_type__": "node",
      },
    },
    "uuid-4": {
      "__NODE__": true,
      "children": {},
      "id": "uuid-4",
      "locked": undefined,
      "parent": "__ROOT__",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "edge",
        "__node_type__": "edge",
        "source": {
          "cell": "uuid-0",
        },
        "target": {
          "cell": "uuid-3",
        },
      },
    },
  },
}
`;

exports[`buildEditorYjs with edges 1`] = `
{
  "test": {
    "__ROOT__": {
      "__NODE__": true,
      "children": {
        "uuid-0": 1,
        "uuid-1": 1,
        "uuid-2": 1,
        "uuid-3": 1,
      },
      "id": "__ROOT__",
      "locked": undefined,
      "parent": undefined,
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "__ROOT__",
        "__node_type__": "node",
      },
    },
    "uuid-0": {
      "__NODE__": true,
      "children": {},
      "id": "uuid-0",
      "locked": undefined,
      "parent": "__ROOT__",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "a",
        "__node_type__": "node",
      },
    },
    "uuid-1": {
      "__NODE__": true,
      "children": {},
      "id": "uuid-1",
      "locked": undefined,
      "parent": "__ROOT__",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "b",
        "__node_type__": "node",
      },
    },
    "uuid-2": {
      "__NODE__": true,
      "children": {},
      "id": "uuid-2",
      "locked": undefined,
      "parent": "__ROOT__",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "my-edge-a",
        "__node_type__": "edge",
        "source": {
          "cell": "uuid-0",
        },
        "target": {
          "cell": "uuid-1",
        },
      },
    },
    "uuid-3": {
      "__NODE__": true,
      "children": {},
      "id": "uuid-3",
      "locked": undefined,
      "parent": "__ROOT__",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "my-edge-b",
        "__node_type__": "edge",
        "source": {
          "cell": "uuid-1",
        },
        "target": {
          "cell": "keep-it",
          "port": "port-a",
        },
      },
    },
  },
}
`;

exports[`buildEditorYjs with nodes 1`] = `
{
  "test": {
    "__ROOT__": {
      "__NODE__": true,
      "children": {
        "keep-it": 1,
        "uuid-0": 1,
      },
      "id": "__ROOT__",
      "locked": undefined,
      "parent": undefined,
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "__ROOT__",
        "__node_type__": "node",
      },
    },
    "keep-it": {
      "__NODE__": true,
      "children": {},
      "id": "keep-it",
      "locked": undefined,
      "parent": "__ROOT__",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "my-node",
        "__node_type__": "node",
        "a": 1,
        "b": 2,
      },
    },
    "uuid-0": {
      "__NODE__": true,
      "children": {},
      "id": "uuid-0",
      "locked": undefined,
      "parent": "__ROOT__",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "foo",
        "__node_type__": "node",
        "b": 1,
        "c": 2,
        "d": "uuid-0",
      },
    },
  },
}
`;

exports[`createRoot 1`] = `
{
  "test": {
    "root": {
      "__NODE__": true,
      "children": {
        "1": 1,
        "2": 1,
        "3": 1,
      },
      "id": "__ROOT__",
      "locked": undefined,
      "parent": undefined,
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "__ROOT__",
        "__node_type__": "node",
      },
    },
  },
}
`;
